src/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.cs

```diff
@@ -0,0 +1,293 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+namespace System.Threading
+{
+    /// <summary>
+    /// A LIFO semaphore.
+    /// Waits on this semaphore are uninterruptible.
+    /// </summary>
+    internal partial class LowLevelLifoSemaphore : IDisposable
+    {
+        private CacheLineSeparatedCounts _separated;
+
+        private int _maximumSignalCount;
+
+        private static int s_processorCount = Environment.ProcessorCount;
+
+        public LowLevelLifoSemaphore(int initialSignalCount, int maximumSignalCount)
+        {
+            Debug.Assert(initialSignalCount >= 0);
+            Debug.Assert(initialSignalCount <= maximumSignalCount);
+
+            _separated = new CacheLineSeparatedCounts();
+            _separated._counts._signalCount = (uint)initialSignalCount;
+            _maximumSignalCount = maximumSignalCount;
+
+            Create(maximumSignalCount);
+        }
+
+        public bool Wait(int timeoutMs)
+        {
+            // Try to acquire the semaphore or
+            // a) register as a spinner if spinCount > 0 and timeoutMs > 0
+            // b) register as a waiter if there's already too many spinners or spinCount == 0 and timeoutMs > 0
+            // c) bail out if timeoutMs == 0 and return false
+            Counts counts = Counts.VolatileRead(ref _separated._counts);
+            while (true)
+            {
+                Debug.Assert(counts._signalCount <= _maximumSignalCount);
+                Counts newCounts = counts;
+
+                if (counts._signalCount != 0)
+                {
+                    newCounts._signalCount--;
+                }
+                else if (timeoutMs != 0)
+                {
+                    if (LowLevelSpinWaiter.SpinCount > 0 && newCounts._spinnerCount < byte.MaxValue)
```

Comment:
The spin count should be passed in so that runtimes may control it. For instance in CoreCLR's implementation it is configurable to allow users to disable spinning there, and on Unixes where the spin count is doubled because of CoreCLR's PAL overhead.

Reply:
I will add it as a parameter. The tuning for "PAL overhead" may not be needed here, but only testing can show that.

Reply:
It's not needed for CoreRT. Depends on the runtime's implementation. CoreCLR's `LowLevelLifoSemaphore` would use the PAL so it would probably need the doubling (as in the issue that resulted in doubling the spin count does not change with the portable thread pool implementation). The actual runtime choice of spin count could be flushed out later, for CoreRT for now it can use the default spin count used in CoreCLR (see `ThreadPool_UnfairSemaphoreSpinLimit` in `clrconfigvalues.h`, default is 70 currently).

Reply:
Maybe it can be made configurable similarly to other config options in the portable thread pool

Reply:
I already added the same `ThreadPool_UnfairSemaphoreSpinLimit` app context switch with the same default value.